lib/fa2/asset/multi_asset.impl.jsligo

```diff
@@ -0,0 +1,263 @@
+         Operators,
+         address,
+         operator,
+         nat
+      ]
+   ): Operators => {
+      if (owner == operator) {
+         return operators
+      } // assert_authorisation always allow the owner so this case is not relevant
+       else {
+         assert_update_permission(owner);
+         let auth_tokens =
+            match(Big_map.find_opt([owner, operator], operators)) {
+               when (Some(ts)):
+                  ts
+               when (None()):
+                  Set.empty
+            };
+         auth_tokens = Set.add(token_id, auth_tokens);
+         return Big_map.update([owner, operator], Some(auth_tokens), operators)
+      }
+   };
+   export const remove_operator = (
+      [operators, owner, operator, token_id]: [
+         Operators,
+         address,
+         operator,
+         nat
+      ]
+   ): Operators => {
+      if (owner == operator) {
+         return operators
+      } // assert_authorisation always allow the owner so this case is not relevant
+       else {
+         assert_update_permission(owner);
+         const auth_tokens: option<set<nat>> =
+            match(Big_map.find_opt([owner, operator], operators)) {
+               when (Some(toks)):
+                  do {
+                     const ts = Set.remove(token_id, toks);
+                     if (Set.cardinal(ts) == 0n) {
+                        return None()
+                     } else {
+                        return Some(ts)
+                     }
+                  }
+               when (None()):
+                  None()
+            };
+         return Big_map.update([owner, operator], auth_tokens, operators)
+      }
+   }
+   //export namespace Ledger {
+
+   export const get_for_user = (
+      [ledger, owner, token_id]: [Ledger, address, nat]
+   ): nat =>
+      match((Big_map.find_opt([owner, token_id], ledger))) {
+         when (Some(a)):
+            a
+         when (None()):
+            0 as nat
+      };
+   const set_for_user = (
+      [ledger, owner, token_id, amount_]: [Ledger, address, nat, nat]
+   ): Ledger =>
+      Big_map.update([owner, token_id], Some(amount_), ledger);
+   export const decrease_token_amount_for_user = (
+      [ledger, from_, token_id, amount_]: [Ledger, address, nat, nat]
+   ): Ledger => {
+      let balance_ = get_for_user([ledger, from_, token_id]);
+      assert_with_error((balance_ >= amount_), Errors.ins_balance);
+      balance_ = abs(balance_ - amount_);
+      return set_for_user([ledger, from_, token_id, balance_])
+   };
+   export const increase_token_amount_for_user = (
+      [ledger, to_, token_id, amount_]: [Ledger, address, nat, nat]
+   ): Ledger => {
+      let balance_ = get_for_user([ledger, to_, token_id]);
+      balance_ = balance_ + amount_;
+      return set_for_user([ledger, to_, token_id, balance_])
+   }
+   //export namespace Storage {
+
+   export const assert_token_exist = (s: storage, token_id: nat): unit => {
```

Comment:
Why take the storage here and not token_metadata(s) directly ? Also should probably be in Assert namespace

Reply:
legacy code